### PR TITLE
#1328 files backend on s3

### PR DESCRIPTION
**Pull request by pazembrz**

### CODE REVIEW

File: hepcrawl/spiders/aps_spider.py [hunk 1]
```diff
@@ -55,9 +54,10 @@ class APSSpider(LastRunStoreSpider):
 
     @strict_kwargs
     def __init__(self, from_date=None, until_date=None,
-                 date="published", sets=None, per_page=100,
+                 date="published", sets=None, per_page=100, aps_url=None,
```
Comment (michamos) [inline hepcrawl/spiders/aps_spider.py:57]: is this for easier testing?

File: hepcrawl/spiders/desy_spider.py [hunk 2]
```diff
@@ -100,27 +90,22 @@ def __init__(
...
-                self.s3_resource = self.get_s3_resource()
+        if not (
+                self.s3_server
+                and self.s3_key
+                and self.s3_secret
+                and self.s3_input_bucket
+                and self.s3_output_bucket
+        ):
+            raise Exception("Missing s3 connection parameters")
+        else:
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py:101]: I think you can remove the `else` here due to the `raise` right above.

File: hepcrawl/spiders/aps_spider.py (hunk 1 above)
Comment (pazembrz) [reply hepcrawl/spiders/aps_spider.py:57]: Yes. Also I hate hard-coded URLs which you cannot change easily.

File: hepcrawl/spiders/desy_spider.py (hunk 2 above)
Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:101]: :+1:

File: hepcrawl/spiders/desy_spider.py [hunk 3]
```diff
@@ -297,22 +258,24 @@ def _parsed_items_from_marcxml(
...
+                            files_to_download.append(document['url'])
+
+                    if new_documents:
+                        parsed_item.record['documents'] = new_documents
 
                     parsed_item.file_urls = files_to_download
-
                     self.logger.info('Got the following attached documents to download: %s' % files_to_download)
-                    self.logger.info('Got item: %s' % parsed_item)
+                    self.logger.info('Got item: %s', parsed_item)
```
Comment (drjova) [inline hepcrawl/spiders/desy_spider.py:278]: why only here :)

File: tests/functional/aps/test_aps.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    yield
+
+    clean_dir(path=os.path.join(os.getcwd(), '.scrapy'))
+
+def get_settings():
+    key = 'key'
+    secret = 'secret'
+    s3_host = 'http://localstack:4566'
+
+
```
Comment (drjova) [inline tests/functional/aps/test_aps.py:33]: too many spaces

File: hepcrawl/settings.py [hunk 5]
```diff
@@ -121,13 +121,16 @@
...
-    'files'
-)
 FILES_URLS_FIELD = 'file_urls'
 FILES_RESULT_FIELD = 'files'
 
+# S3 Settings
+DOWNLOAD_BUCKET = os.environ.get("APP_DOWNLOAD_BUCKET", "documents")
+AWS_ENDPOINT_URL = os.environ.get("APP_AWS_ENDPOINT_URL", "https://s3.cern.ch")
+FILES_STORE = "s3://{bucket}/".format(bucket=DOWNLOAD_BUCKET)
+AWS_ACCESS_KEY_ID = os.environ.get("APP_AWS_ACCESS_KEY_ID")
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply hepcrawl/settings.py:131]: Required by scrapy if we use pipeline with s3 storage enabled: https://docs.scrapy.org/en/latest/topics/media-pipeline.html#amazon-s3-storage

File: tests/functional/desy/test_desy.py [hunk 6]
```diff
@@ -66,125 +70,92 @@ def _override(field_key, original_dict, backup_dict, new_value):
...
+    )
+    transfer_config = TransferConfig(use_threads=False)
+    for bucket_name, file_name in files_to_upload:
+        buckets_map[bucket_name].upload_file(
+            Filename=os.path.join(test_files_path, file_name),
+            Key=file_name, Config=transfer_config
+        )
 
-    bucket.create()
+def establish_s3_connection(s3_key, s3_secret, s3_server):
```
Comment (drjova) [inline tests/functional/desy/test_desy.py:93]:
```suggestion
def s3_connection(s3_key, s3_secret, s3_server):
```

File: hepcrawl/spiders/desy_spider.py (hunk 3 above)
Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:278]: I noticed it only here. :P I will look for more.

File: tests/functional/aps/test_aps.py (hunk 4 above)
Comment (pazembrz) [reply tests/functional/aps/test_aps.py:33]: :+1:

File: hepcrawl/spiders/elsevier_spider.py [hunk 7]
```diff
@@ -203,9 +203,17 @@ def unzip_zip_package_to_s3(self, response):
                         callback=self.parse_record,
                     )
 
+    def _file_name_from_url(self, url):
+        url = url.split("?")[0]
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply hepcrawl/spiders/elsevier_spider.py:207]: :+1:

File: hepcrawl/spiders/desy_spider.py (hunk 3 above)
Comment (drjova) [reply hepcrawl/spiders/desy_spider.py:278]: Does it matter :)? I mean just leave it as is, anyway we have it everywhere.

File: tests/functional/desy/test_desy.py (hunk 6 above)
Comment (pazembrz) [reply tests/functional/desy/test_desy.py:93]: :+1:

File: hepcrawl/spiders/desy_spider.py (hunk 3 above)
Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:278]: In logger module it's recommended to use it's own formatter.

File: hepcrawl/spiders/desy_spider.py [hunk 8]
```diff
@@ -100,27 +90,22 @@ def __init__(
...
+                and self.s3_input_bucket
+                and self.s3_output_bucket
+        ):
+            raise Exception("Missing s3 connection parameters")
+
+        self.s3_connections = self.connect_to_s3()
+        self.s3_resource = self.get_s3_resource()
 
         if not os.path.exists(self.destination_folder):
             os.makedirs(self.destination_folder)
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py:106]: is this still used?

File: hepcrawl/spiders/aps_spider.py [hunk 9]
```diff
@@ -55,9 +54,10 @@ class APSSpider(LastRunStoreSpider):
...
                  **kwargs):
         """Construct APS spider."""
+        self.aps_url = aps_url or self.aps_base_url
```
Comment (michamos) [inline hepcrawl/spiders/aps_spider.py:60]: maybe remove `aps_base_url` and put it as default for the `aps_url` arg?

File: hepcrawl/spiders/desy_spider.py [hunk 10]
```diff
@@ -192,82 +168,66 @@ def crawl_s3_bucket(self):
...
-            parsed_url = urllib.parse.urlparse(url)
-            return not parsed_url.scheme
+    @classmethod
+    def _is_local_path(cls, url):
+        parsed_url = urllib.parse.urlparse(url)
+        return not parsed_url.scheme
 
-        return _is_local_path(current_url)
+    @classmethod
+    def _has_to_be_downloaded(cls, current_url):
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py:183]: what's the point of this? `_has_to_be_downloaded` is the same function as `_is_local_path`. Do we really need both?

File: hepcrawl/spiders/elsevier_spider.py [hunk 11]
```diff
@@ -203,9 +204,15 @@ def unzip_zip_package_to_s3(self, response):
...
+    def _file_name_from_url(self, url):
+        return basename(urlparse(url).path)
+
     def parse_record(self, response):
         """Parse an elsevier XML downloaded from s3 into a HEP record."""
         parser = ElsevierParser(response.meta['data'])
-
-        return ParsedItem(record=parser.parse(), record_format="hep")
+        file_name = self._file_name_from_url(response.url)
+        parser.attach_fulltext_document(file_name, response.url)
```
Comment (michamos) [inline hepcrawl/spiders/elsevier_spider.py:214]: not sure it's happening, but the attach fulltext should be hidden

File: hepcrawl/spiders/aps_spider.py [hunk 12]
```diff
@@ -124,10 +123,12 @@ def _parse_jats(self, response):
 
         file_name = self._file_name_from_url(response.url)
         parser.attach_fulltext_document(file_name, response.url)
```
Comment (michamos) [inline hepcrawl/spiders/aps_spider.py:125]: I don't know if it's happening, but the attached fulltext should be hidden

File: hepcrawl/spiders/desy_spider.py (hunk 8 above)
Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:106]: Yes. This is different s3 stuff :) We have 1 for DESY itself to get data, and there is second one in Pipeline (managed by scrapy) to upload documents. They are completely independent (unfortunately).

Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:106]: O, you are asking about destination folder.... Sorry I looked automatically on "green" stuff ;)
Yeah, this is not necessary. :+1:

File: hepcrawl/spiders/aps_spider.py (hunk 9 above)
Comment (pazembrz) [reply hepcrawl/spiders/aps_spider.py:60]: :+1:

File: hepcrawl/spiders/desy_spider.py (hunk 10 above)
Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:183]: Actually I forgot to remove it, it's not used anumore. Thanks :+1:

File: hepcrawl/spiders/elsevier_spider.py (hunk 11 above)
Comment (pazembrz) [reply hepcrawl/spiders/elsevier_spider.py:214]: Yes it is. Here `attach_fulltext_document` is used and as you can se in  it's [definition](https://github.com/inspirehep/hepcrawl/blob/52db9383d077c24b726137128428abff0a20af38/hepcrawl/parsers/elsevier.py#L606-L607) it's adding file as hidden.

File: hepcrawl/spiders/aps_spider.py (hunk 12 above)
Comment (pazembrz) [reply hepcrawl/spiders/aps_spider.py:125]: Same here :) https://github.com/inspirehep/hepcrawl/blob/52db9383d077c24b726137128428abff0a20af38/hepcrawl/parsers/jats.py#L588-L589

File: hepcrawl/spiders/desy_spider.py [hunk 13]
```diff
@@ -162,23 +142,10 @@ def _filter_xml_files(list_files_paths):
...
-            )
-            yield Request(
-                'file://{0}'.format(file_path),
-                callback=self.parse,
-            )
 
     def crawl_s3_bucket(self):
         input_bucket = self.s3_resource.Bucket(self.s3_input_bucket)
-        existing_files = os.listdir(self.destination_folder)
+        existing_files = [file.key for file in self.s3_resource.Bucket(self.s3_output_bucket).objects.all()]
```
Comment (michamos) [inline hepcrawl/spiders/desy_spider.py:148]: this is going to be expensive if the bucket is a bit full, better skip this check, or check it for every file through the API if really needed.

Comment (pazembrz) [reply hepcrawl/spiders/desy_spider.py:148]: Yes it is needed (It was in original - FTP - implementation) to be sure that we are not processing same file twice. I will add check through API then.

**Review by michamos — APPROVED** [review]
I'm worried about the "archaeological" complexity (different layers with different ages have a completely different style and logic) of this whole project, making this kind of changes *extremely* hard to review. I don't have any concrete suggestions for quick improvement (I think you did a good job @pazembrz given the constraints). It doesn't look completely crazy and seems well-tested, so let's merge this.

**Pull request merged**